narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,184 @@
+use config::{AuthorityIdentifier, Committee};
+use crypto::RandomnessPrivateKey;
+use fastcrypto::groups;
+use fastcrypto_tbls::{dkg, nodes};
 use mysten_metrics::metered_channel::{Receiver, Sender};
 use mysten_metrics::spawn_logged_monitored_task;
+use sui_protocol_config::ProtocolConfig;
 use tap::TapFallible;
 use tokio::task::JoinHandle;
 use tracing::{debug, error, info, warn};
-use types::{Certificate, CertificateAPI, ConditionalBroadcastReceiver, HeaderAPI, Round};
+use types::{
+    Certificate, CertificateAPI, ConditionalBroadcastReceiver, HeaderAPI, Round, SystemMessage,
+};
 
-/// Receives the highest round reached by consensus and update it for all tasks.
+type PkG = groups::bls12381::G2Element;
+type EncG = groups::bls12381::G2Element;
+
+/// Updates Narwhal system state based on certificates received from consensus.
 pub struct StateHandler {
-    /// The id of this authority.
     authority_id: AuthorityIdentifier,
+
     /// Receives the ordered certificates from consensus.
     rx_committed_certificates: Receiver<(Round, Vec<Certificate>)>,
     /// Channel to signal committee changes.
     rx_shutdown: ConditionalBroadcastReceiver,
     /// A channel to update the committed rounds
     tx_committed_own_headers: Option<Sender<(Round, Vec<Round>)>>,
+    /// A channel to send system messages to the proposer.
+    tx_system_messages: Sender<SystemMessage>,
+
+    /// If set, generates Narwhal system messages for random beacn
+    /// DKG and randomness generation.
+    randomness_state: Option<RandomnessState>,
 
     network: anemo::Network,
 }
 
+// Internal state for randomness DKG and generation.
+struct RandomnessState {
+    party: dkg::Party<PkG, EncG>,
+    messages: Vec<dkg::Message<PkG, EncG>>,
+    processed_messages: Vec<dkg::ProcessedMessage<PkG, EncG>>,
+    shares: dkg::SharesMap<<PkG as groups::GroupElement>::ScalarType>,
+    confirmations: Vec<dkg::Confirmation<EncG>>,
+    dkg_output: Option<dkg::Output<PkG, EncG>>,
+}
+
+impl RandomnessState {
+    fn try_new(
+        protocol_config: ProtocolConfig,
+        committee: Committee,
+        private_key: RandomnessPrivateKey,
+    ) -> Option<Self> {
+        if !protocol_config.random_beacon() {
+            return None;
+        }
+
+        let info = committee.randomness_dkg_info();
+        let nodes = info
+            .iter()
+            .map(|(id, pk, stake)| nodes::Node::<EncG> {
+                id: id.0,
+                pk: pk.clone(),
+                weight: *stake as u16,
+            })
+            .collect();
+        let nodes = match nodes::Nodes::new(nodes) {
+            Ok(nodes) => nodes,
+            Err(err) => {
+                error!("Error while initializing random beacon state: {err:?}");
+                return None;
+            }
+        };
+        // TODO_DNS do we expect to want to vary this in the future? in which case it should be in protocl config
+        const DKG_THRESHOLD: u16 = 3_334; // f+1 of total 10,000 stake.
```

Comment:
this is available as VALIDITY_THRESHOLD in sui-types/src/committee.rs

i forget if narwhal uses any sui-types code, but if not you may want to move this to the narwhal committee module.